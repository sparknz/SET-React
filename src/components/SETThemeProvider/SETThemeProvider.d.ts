/// <reference types="react" />
export default function SparkThemeProvider({ children }: {
    children: any;
}): JSX.Element;
