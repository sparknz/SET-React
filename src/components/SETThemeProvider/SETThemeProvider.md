### Usage

The `SETThemeProvider` provide styled-components in your application with the color/layout/sizing tokens for the specific chosen theme.

You usually only need to use the `SETThemeProvider` once, and that is at the root element of your project:

```jsx static
class App extends Page {
    render() {
        return (
            <SparkTheme>
                {/*Your app goes here*/}
            </SparkTheme>
        );
    }
}
```

### Helper functions
The easiest way to use these tokens is via the theme helper functions

#### Color helpers
There are 7 different color helper functions:
- backgroundColor
- textColor
- borderColor
- borderTopColor
- borderRightColor
- borderBottomColor
- borderLeftColor

> Note: you should never need to use the backgroundColor helper file directly, as any div with a background should be created using the Box component.

The color helpers can be used like so:
``` jsx static
const MyComponent1 = styled.p`
    ${textColor('default')} //color: #123456;
`;

const MyComponent2 = styled.span`
    ${borderColor('alt1')} //border-color: #123456;
`;

const MyComponent3 = styled.div`
    ${backgroundColor('error')} //border-bottom: #123456;
`;
```

#### Layout helpers
There are 2 different layout helper functions:
- breakpoint
- zIndex


You could use the breakpoint helper like so:
```jsx static
import styled from 'styled-components';
import { textColor, breakpoint, zIndex } from '../helpers/themeHelpers';

export default styled.h1`
    ${textColor('success')}
    ${breakpoint('tablet')}{
        ${textColor('error')}
    }
`;
```

And here is the result:

```jsx noeditor
import MediaQuery from '../../demo-components/MediaQuery';

<MediaQuery>This is a responsive text, change the window size to see me change color</MediaQuery>
```

This is how you use the zIndex helper:
```jsx static
import React from 'react';
import styled from 'styled-components';
import { textColor, breakpoint, zIndex } from '../helpers/themeHelpers';

const Lower = styled.h1`
    position: absolute;
    ${zIndex('bottomlessPit')}
`;

const Higher = styled.h4`
    position: absolute;
    ${zIndex('overTheMoon')}
`

const Wrapper = styled.div`
    position: relative;
`

export default function(){
    return <Wrapper>
        <Lower>Text in bottomless pit</Lower>
        <Higher>Text that's over the moon</Higher>
    </Wrapper>
}
```

#### Sizing helpers


### Raw Theme Object


Refer to [Spark Tokens](https://sparknz.github.io/SET-Docs) for all the avaliable colors, layout, and spacing values 