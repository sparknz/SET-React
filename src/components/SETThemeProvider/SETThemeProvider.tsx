/* eslint-disable no-param-reassign */

import React from 'react';
import tokens from '@sparknz/set-tokens/build/web/tokensWithVariations';
import {
    ThemeProvider, css, FlattenInterpolation, ThemeProps,
} from 'styled-components';
import merge from 'deepmerge';
import { prop } from 'styled-tools';
import { Colors } from './interfaces';

type CSS = FlattenInterpolation<ThemeProps<any>>;

function createColorGetter(property: string): Colors {
    const res = {};
    const rec = (thing, otherThing, path): void => {
        Object.keys(thing).forEach((key): void => {
            if (thing[key].value) {
                otherThing[key] = (props): CSS => {
                    if (thing[key].warning) {
                        console.warn(thing[key].warning);
                    }
                    return css`${property}${prop(`theme.color${path}.${key}.value`)(props)}`;
                };
            } else {
                if (!otherThing[key]) {
                    otherThing[key] = {};
                }
                rec(thing[key], otherThing[key], `${path}.${key}`);
            }
        });
    };
    rec(tokens.base.color, res, '');
    return res as Colors;
}

export const backgroundColor: Colors = createColorGetter('background-color: ');
export const textColor: Colors = createColorGetter('color: ');
export const borderColor: Colors = createColorGetter('border-color: ');
export const borderColorLeft: Colors = createColorGetter('border-left-color: ');
export const borderColorRight: Colors = createColorGetter('border-right-color: ');
export const borderColorTop: Colors = createColorGetter('border-top-color: ');
export const borderColorBottom: Colors = createColorGetter('border-bottom-color: '); 

export default ({ theme = 'purple-orange', ...props }) => {
    const constructedTheme = merge(tokens.base, tokens.themes[theme]);
    return <ThemeProvider {...props} theme={constructedTheme} />;
};
