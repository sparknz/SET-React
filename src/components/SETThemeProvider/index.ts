export { 
    default,
    backgroundColor,
    textColor,
    borderColor,
    borderColorLeft,
    borderColorRight,
    borderColorTop,
    borderColorBottom,
} from './SETThemeProvider';