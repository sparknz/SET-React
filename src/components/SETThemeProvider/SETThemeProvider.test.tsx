import React from 'react';
import renderer from 'react-test-renderer';
import styled from 'styled-components';
import 'jest-styled-components';

import SETThemeProvider, {
    backgroundColor,
    textColor,
    borderColor, 
    borderColorLeft,
    borderColorRight, 
    borderColorTop,
    borderColorBottom 
} from './index';

console.warn = jest.fn();

test('It renders without crashing', () => {
    const component = renderer.create(<SETThemeProvider />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
})


const TestStyle = styled.div`
    ${backgroundColor.base.white};
    ${textColor.base.primary.base};
    ${borderColor.base.primary.light};
    ${borderColorLeft.base.secondary.light};
    ${borderColorRight.base.secondary.base};
    ${borderColorTop.base.neutral['000']};
    ${borderColorBottom.base.tertiary};

`;

test('Background.base.white returns the correct color', () => {
    const component = renderer.create(<SETThemeProvider>
        <TestStyle />
    </SETThemeProvider>);
    let tree = component.toJSON();
    expect(tree).toHaveStyleRule('color', '#5f259f');
    expect(tree).toHaveStyleRule('background-color', '#ffffff');
    expect(tree).toHaveStyleRule('border-color', '#75529d');
    expect(tree).toHaveStyleRule('border-left-color', '#f9b104');
    expect(tree).toHaveStyleRule('border-right-color', '#ff9b00');
    expect(tree).toHaveStyleRule('border-bottom-color', '#03a9f4');
    expect(tree).toHaveStyleRule('border-top-color', '#f8f9f9');
})


const TestBrandColor = styled.div`
    ${backgroundColor.brand.green.base};
`;


test('Background.brand.green.base returns the correct color and calls a warning', () => {
    expect(console.warn).not.toBeCalled();
    const component = renderer.create(<SETThemeProvider>
        <TestBrandColor />
    </SETThemeProvider>);
    let tree = component.toJSON();
    expect(tree).toHaveStyleRule('background-color', '#00af55');
    expect(console.warn).toBeCalled();
})
