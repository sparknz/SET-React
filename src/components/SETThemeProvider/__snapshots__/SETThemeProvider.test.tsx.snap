// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It renders without crashing 1`] = `null`;
